crates/cpuid-utils/src/lib.rs

```diff
@@ -23,55 +63,494 @@ pub mod bits;
 #[cfg(feature = "instance-spec")]
 mod instance_spec;
 
-#[cfg(feature = "instance-spec")]
-pub use instance_spec::*;
+type CpuidSubleafMap = BTreeMap<u32, CpuidValues>;
+type CpuidMapInsertResult = Result<Option<CpuidValues>, SubleafInsertConflict>;
 
-/// A map from CPUID leaves to CPUID values.
+/// Denotes the presence or absence of subleaves for a given CPUID leaf.
+#[derive(Clone, Debug, PartialEq, Eq)]
+enum Subleaves {
+    /// This leaf is configured to have subleaves, whose values are stored in
+    /// the inner map.
+    Present(CpuidSubleafMap),
+
+    /// This leaf does not have subleaves.
+    Absent(CpuidValues),
+}
+
+/// [`CpuidMap`]'s insert functions return this error if a request to insert a
+/// new value would produce a leaf that has both per-subleaf values and a
+/// no-subleaf value.
+#[derive(Debug, Error)]
+pub enum SubleafInsertConflict {
+    #[error("leaf {0:x} has entries with subleaves")]
+    SubleavesAlreadyPresent(u32),
+
+    #[error("leaf {0:x} has an entry with no subleaves")]
+    SubleavesAlreadyAbsent(u32),
+}
+
+/// A mapping from CPUID leaf/subleaf pairs to CPUID return values. This struct
+/// allows each registered leaf either to have or not to have subleaf values,
+/// but not both at once.
+///
+/// Some CPUID leaves completely ignore the subleaf value passed in ecx. Others
+/// pay attention to it and have their own per-leaf semantics that govern what
+/// happens if the supplied ecx value is out of the function's expected range.
+/// Tracking CPUID values in a simple `BTreeMap` from [`CpuidIdent`] to
```

Comment:
as we found out a bit earlier, it's also acceptable (if not well-defined) for callers to omit subleaves, leaving a hole in the range `0...max_ecx`. if not documented here, it feels like that's worth mentioning somewhere around this type... maybe just a variant in the proptest that asserts such a construction is OK?

Reply:
Good call--amended the doc comment in 80b3cc6.